Rainforests training plugins

lukehoffmann opened the pull request:
Description

Adds some plugins to perform the basic tasks for training RainForests calibration models.

- `TrainRainForestsModel` to generate the LightGBM tree, and
- `CompileRainForestsModel` to compile these with Treelite

Testing:

- [x] Ran tests and they passed OK
- [x] Added new tests for the new feature(s)

CLA

- [x] If a new developer, signed up to CLA

lukehoffmann committed:
Refactor module checks into common area

lukehoffmann committed:
Plugins for training and compiling Rainforests models

lukehoffmann committed:
Add some testing for column selection

lukehoffmann committed:
Standardise some naming and documentation

lukehoffmann committed:
Fix import

lukehoffmann committed:
Bump up test coverage

benowen-bom reviewed (commented):
I've had a brief look over your PR. I think this is a really good place to build up from and the logic you have there is sound.

I have some broad comments which would be good to put up front before committing to a detailed review. The first is a suggestion, and the other question, both relating to structural considerations which would be good to pin down now.
 
**Suggestion:**
the calibration works by building a series of models defined over a set of thresholds. I would suggest that threshold argument in process function within TrainRainForestsModel become a list of thresholds thresholds and add a loop over these thresholds to write out a series of files. I appreciate that producing a series of files is not the IMPROVER norm, but ideally we get a series of models to map out the CDF across multiple values

**Question:**
do you think there is any value in bringing the compilation within the TrainRainForestsModel as add a flag to the process function to compile? That would simplify the issue of having to run the compilation over many files and would ensure there is a 1-1 mapping for the Light-GBM files and the compiled treelite files.

rachaelesler commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,73 @@
+                Value is the number of processes to use.
+                Defaults to 0 (no parallel compilation)
+        """
+
```
Suggest not using a string for the default argument for `toolchain` ([explanation why here](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil)). A better approach is to set the default to `None` and handle it in the function: 


```suggestion
    def __init__(self, toolchain=None, verbose=False, parallel_comp=0):
        """Initialise the options used when compiling models.

        Args:
            toolchain:
                Toolchain to use for Treelite model compilation. 
                'gcc', 'msvc', 'clang' or a specific variation of clang or gcc
                (e.g. 'gcc-7'). If not provided, defaults to 'gcc'.
            verbose (bool):
                Print verbose output during compilation
            parallel_comp (int):
                Enables parallel compilation to reduce time and memory consumption.
                Value is the number of processes to use.
                Defaults to 0 (no parallel compilation)
        """
        if toolchain is None: 
            toolchain = "gcc" 
```

lukehoffmann committed:
Accept a list of thresholds for training

lukehoffmann committed:
formatting

lukehoffmann committed:
Add option to compile while training model

lukehoffmann committed:
Refactor training plugin

lukehoffmann commented:
> I've had a brief look over your PR. I think this is a really good place to build up from and the logic you have there is sound.

Thanks @benowen-bom, I've finally got back to this now.

> **Suggestion:** the calibration works by building a series of models defined over a set of thresholds. I would suggest that threshold argument in process function within TrainRainForestsModel become a list of thresholds thresholds and add a loop over these thresholds to write out a series of files. I appreciate that producing a series of files is not the IMPROVER norm, but ideally we get a series of models to map out the CDF across multiple values

That was easy enough to change and looks okay. I've just had to add the threshold to each filename to make them unique.

> **Question:** do you think there is any value in bringing the compilation within the TrainRainForestsModel as add a flag to the process function to compile? That would simplify the issue of having to run the compilation over many files and would ensure there is a 1-1 mapping for the Light-GBM files and the compiled treelite files.

It still makes sense to me to keep a dedicated class for compilation. ~I've tried adding an option to pass this object into the training class and optionally compile at the same time as training. Let me know what you think.~ 
_Update: I've ended up removing this option to train and compile in one step.. It didn't make sense to me when I went through the worked example in [Forecast-Improvement/raincal_lite](https://github.com/Forecast-Improvement/raincal_lite)._

lukehoffmann committed:
Add optional lightgbm parameters

lukehoffmann committed:
Exclude target column from training data

lukehoffmann committed:
Allow lambda to format model filenames

lukehoffmann committed:
Use rainforests config dict in training plugin

This is simpler than generating file paths within the plugin

lukehoffmann committed:
Use rainforests config dict in compiler plugin

lukehoffmann committed:
Fix some types and errors in training plugin

lukehoffmann committed:
Simplify test data setup

lukehoffmann committed:
Add tests for compilation with missing models

benowen-bom commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,139 @@
+            verbose (bool):
+                Print verbose output during compilation
+            parallel_comp (int):
+                Enables parallel compilation to reduce time and memory consumption.
```
Do you know how the parallel computation reduces memory consumption? Naïvely I would have thought it would increase memory consumption (even if just modestly).

lukehoffmann replied:
This text came from the tl2cgen documentation for that option
https://tl2cgen.readthedocs.io/en/latest/compiler_param.html#_CPPv413parallel_comp
> Set this option to improve compilation time and reduce memory consumption during compilation.

If you run the compiler without this option, you get this warning
```
compiler/ast/split.cc:30: Parallel compilation disabled; all member trees will be dumped to a single source file. This may increase compilation time and memory usage.
```

So I this option means it creates multiple C source files and compiles them in parallel. I would also think that could allow it to use more memory to compile everything more quickly. My quick testing seems to back that up, but I probably don't have a big enough model to test with.

I could just remove the mention of memory here.

benowen-bom replied:
No that's fine, that explanation makes sense. Leave as is!

benowen-bom commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,139 @@
+
+            Args:
+                allow_missing (bool):
+                    If False (default), throws if any LightGBM models are missing.
```
```suggestion
                    If False (default), throws an error if any LightGBM models are missing.
```

lukehoffmann replied:
Done

benowen-bom commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,139 @@
+    def process(self, allow_missing=False):
+        """Compile all configured LightGBM models with Treelite.
+
+            Args:
```
Doc-string alignment seems slightly off here.

lukehoffmann replied:
Fixed

benowen-bom commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,139 @@
+        if output_path.suffix.lower() != TREELITE_EXTENSION:
+            raise ValueError(f"Output path must have extension {TREELITE_EXTENSION}")
+
+        Path.mkdir(output_path.parent, parents=True, exist_ok=True)
```
I'm not sure whether this should occur here, or whether it should be assumed that the dir exists; generally directory creation sits outside of improver (ie handled by the suite). That being said, the paradigm of rainforests is somewhat exceptional, and this approach simplifies things significantly, particularly when files across lead-times or threshold reside in separate folders.

Leave this as is, but I would be interested to hear what other think.

benowen-bom commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,144 @@
+
+        Args:
+            model_config_dict:
+                Dictionary containing Rainforests model configuration variables.
```
I feel like that this definition could be clearer. I think it is good to have the fromat summary below, but many this could be replaced with something like:
```suggestion
                Dictionary describing the high-level RainForests model structure;
                - top level key describes the lead-hour, 
                - next level key describes the threshold,
                - corresponding values the associated model file.
```

lukehoffmann replied:
Good suggestion, I've used your version

benowen-bom commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,144 @@
+
+        # Check all specified columns exist in the data.
+        for col in training_columns:
+            if col not in training_data:
```
My pandas is a bit sketchy (been a while). Does this check that `col` is in `training_data.columns`? Naively I would have thought this was checking whether the value `col` was in the data itself (ie, the underlying array).

lukehoffmann replied:
That's a good question - I had to look it up. If you treat a DataFrame like a collection, it will iterate over the "info axis", which seems to mean it will iterate over columns by default.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.__iter__.html#pandas.DataFrame.__iter__

So this is the same as using `data.columns`. I'll change it to use `columns` explicitly, since that would be clearer and potentially safer.

lukehoffmann replied:
I've changed this to `training_data.columns`

benowen-bom commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,144 @@
+        for col in training_columns:
+            if col not in training_data:
+                raise KeyError(f"Training column '{col}' not found in training data.")
+        if observation_column not in training_data:
```
As above.

lukehoffmann replied:
As above, I changed this to `training_data.columns`

benowen-bom commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,144 @@
+
+        # Merge default params with optional params.
+        lightgbm_params = lightgbm_params or {}
+        self.lightgbm_params = self.lightgbm_params | lightgbm_params
```
I have a few questions here:
- seeing as `lightgmb_params` is an arg to the `__init__` function, how is the class level variable `lightgbm_params` used? I would imagine that it is overwritten in the namespace within this function and never referenced elsewhere. Is that correct? Or is the `self.lightgbm_params` referencing the value defined at the class level?
- I've never seen dictionary handling like this: `lightgbm_params or {}`. How does this work? What is returned in the case `lightgbm_params` dict is defined and when it is `None`?
- How does the dictionary merge behave when common values are across both dictionaries? Does this fail, or is one overriding the other? If overriding, which one takes precedence. I would argue that we would want the `lightgbm_params` passed here to be the one that takes precedence.

lukehoffmann replied:
> I have a few questions here:
> 
> * seeing as `lightgmb_params` is an arg to the `__init__` function, how is the class level variable `lightgbm_params` used? I would imagine that it is overwritten in the namespace within this function and never referenced elsewhere. Is that correct? Or is the `self.lightgbm_params` referencing the value defined at the class level?

I think I'll rework this a bit to make it more obvious what's happening. I'll avoid using the same name for variables/members in different scopes.

> * I've never seen dictionary handling like this: `lightgbm_params or {}`. How does this work? What is returned in the case `lightgbm_params` dict is defined and when it is `None`?

This isn't specific to dictionaries, it's just a shorthand way of initialising a variable if it may or may not have a value yet. It's taking advantage of the fact that the `or` operator doesn't necessarily return a Boolean, it just returns the first operand if that first operand equates to `True`, otherwise it returns the second operand. I've seen this a lot in other languages but maybe it's not too common in Python. There is a proposal for a "null coalescing operator" which would do this a bit more safely. This would be the equivalent with a Python ternary.

```
lightgbm_params = lightgbm_params if lightgbm_params is not None else {}
```

> * How does the dictionary merge behave when common values are across both dictionaries? Does this fail, or is one overriding the other? If overriding, which one takes precedence. I would argue that we would want the `lightgbm_params` passed here to be the one that takes precedence.

This is the [dict union operator, added in Python 3.9](https://docs.python.org/3/whatsnew/3.9.html#dictionary-merge-update-operators). It will combine the two dicts, and for any clashing keys it will take the value from the second dict. So yes the idea is that the params passed into the constructor will take precedence over the default params.

lukehoffmann replied:
I renamed the class member to just `params` to avoid the ambiguity with the class member.

I've removed this initialisation:
```
lightgbm_params = lightgbm_params or {}
```
Instead I've just done `if lightgbm_params:` before merging the two dictionaries.

I've added some unit tests to make sure these parameters are handled the way I expect, and also to check that they're actually used in the lightgbm training.

benowen-bom replied:
Thanks for updating this, and the explanation above. With the changes to the code and the updated comment, I think this is much clearer now.

benowen-bom reviewed (commented):
Sorry, this is only a partial review: I'll review the tests next, but want to make sure I don't lose comments so far.

lukehoffmann committed:
Improve readability from review comments

benowen-bom commented on improver_tests/calibration/rainforests_training/conftest.py:
```diff
@@ -0,0 +1,111 @@
+
+
+@pytest.fixture
+def model_config(tmp_path):
```
Correct me if I'm wrong, but is it best practice to avoid having unit-tests write data to disk? That being said, I think the usage here constitutes a case where it would be warranted given how the treelite compilation works. So I have two questions:
1. where is the data written to disk? (ie what is `tmp_path`), and
2. is there any housekeeping once the tests have executed?

lukehoffmann replied:
The `tmp_path` fixture is provided by pytest. https://docs.pytest.org/en/stable/how-to/tmp_path.html#temporary-directory-location-and-retention

1. By default it's going to `/tmp/pytest-of-lhoffman/pytest-*` on Gadi.
2. Pytest seems to be doing a good job of cleaning up old directories. I only have a few directories there, all from today.

lukehoffmann replied:
Also this `tmp_path` fixture is used pretty widely in IMPROVER tests, and I can only see one or two cases where it is explicitly cleaned up.

benowen-bom replied:
No worries at all. I haven't come across this in my work, so I thought it worth checking. It seems that this is the write way to do this, so I'll keep this in mind in case I ever need to do something similar!

benowen-bom commented on improver_tests/calibration/rainforests_training/test_CompileRainForestsModel.py:
```diff
@@ -0,0 +1,97 @@
+    for lead_time, thresholds in model_config.items():
+        for threshold in thresholds:
+            path = model_config[lead_time][threshold]["treelite_model"]
+            assert Path(path).exists()
```
This is probably overkill given we `process` file is creating treelite files, but would it also be worthwhile loading the written files to assert that they are valid treelite model objects?

lukehoffmann replied:
That's a good idea. I've added a step within this loop to load the model with tl2cgen. There's not much you can inspect on the object without actually evaluating it. There was a property for number of features, so I added a check for that.

benowen-bom commented on improver_tests/calibration/rainforests_training/test_TrainRainForestsModel.py:
```diff
@@ -0,0 +1,141 @@
+            )
+
+
+def test__init__(model_config, deterministic_training_data):
```
I'd suggest adding test for the values of `lightgbm_params`, both default values and when values are passed to the `__init__` function.

benowen-bom replied:
I think it would also be useful to check that `traning_data` is just comprised of the columns `training_columns + [observation_column]`.

lukehoffmann replied:
I agree with the first comment, I've added the tests:
* `test_process_calls_train_with_default_params`
* `test_process_calls_train_with_override_params`

For the second comment, I don't think there's any problem with including unnecessary columns to the constructor of the plugin. The actual lightgbm dataset is reconstructed using only the training columns, and the obs column is used separately to create a label column. Anything else is ignored.

I've added a test to make sure that the actual lightgbm dataset contains the right columns.
* `test_process_calls_train_with_all_columns`

benowen-bom replied:
No worries. Looking over the tests, I'm happy with the updates.

benowen-bom commented on improver_tests/calibration/rainforests_training/test_TrainRainForestsModel.py:
```diff
@@ -0,0 +1,141 @@
+
+    for threshold in thresholds:
+        expected_path = model_config[lead_time][threshold]["lightgbm_model"]
+        assert Path(expected_path).exists()
```
As was mentioned in the treelite case, would it be worth loading in the file to ensure that it is a valid model file?

lukehoffmann replied:
That's a good idea and pretty easy to add, so I've put the check within this loop

benowen-bom commented on improver_tests/calibration/rainforests_training/test_TrainRainForestsModel.py:
```diff
@@ -0,0 +1,141 @@
+        trainer.process(invalid_lead_time, thresholds)
+
+
+def test_process_missing_threshold(model_config, deterministic_training_data):
+    """Test lightgbm models are not created for invalid threshold."""
+
+    lead_time, training_data, observation_column, training_columns = (
+        deterministic_training_data
+    )
+
+    trainer = TrainRainForestsModel(
+        model_config, training_data, observation_column, training_columns
+    )
+
+    invalid_thresholds = list(model_config[lead_time].keys()) + ["1234.5"]
+
+    with pytest.raises(KeyError):
+        trainer.process(lead_time, invalid_thresholds)
```
In the treelite case, the check is done before compiling meaning that we avoid the situation of having partial execution of the `process` function. Do you think we should do something similar here?

lukehoffmann replied:
Good suggestion. I've changed the `process` function to check all thresholds first. I've updated this test to assert that no models are trained in that case.

benowen-bom commented on improver_tests/calibration/rainforests_training/test_TrainRainForestsModel.py:
I wonder if there is a way we could test the `_train_model` through a highly contrived set of inputs that would allow us to test the outputs from the model. I appreciate that this is beyond the task set before you, so it might be worth seeking input from others (myself included).

What I have in mind is that we define some fixed arrays for the input, and create model so simple that we can produce a set of expected values manually to test the preparation of inputs to `lightgbm.train` is being done properly. This would be about the best we could do to test the actual values from the model are sensible.

lukehoffmann replied:
I had a look at the idea of creating some extremely simple test data as a fixture. I would definitely need some more input to decide exactly what to do there. I don't know how much we could simplify the training data and still make it a reasonable test of the RainForests functionality (as opposed to just a test of lightGBM functionality).

I think this is a decent idea, but I would prefer to leave it for a separate issue to do after this PR.

benowen-bom replied:
All good, I appreciate that this is quite hard to do well. I had a look at the code coverage and this section of the code wasn't flagged, so I think I'm being overly pedantic here. Given we are relying on underlying libraries and that you have done extensive testing to produce consistent outputs against what we get from [raincal](https://github.com/metoppv/raincal), I'm happy to abandon this.

Honestly, I don't think this requires an issue unless you think this is really something that _should_ be done.

benowen-bom reviewed (commented):
I've finally finished reviewing the rest of the PR. You've done a remarkable job at taking what was in [raincal](https://github.com/metoppv/raincal) and producing something that is generic and flexible for general calibration problems.

I've got a few comments that would be worth considering. Once we've worked through these I'm happy to approve, but I think it would be worthwhile getting someone else (more technically savvy) to review.

lukehoffmann committed:
Add tests for training parameter overrides

lukehoffmann committed:
Don't train models if any thresholds invalid

lukehoffmann committed:
Add test for training data columns

lukehoffmann committed:
Test loading models after training and compilation

lukehoffmann committed:
Adjust naming of lightgbm param variables

lukehoffmann committed:
Try adding another email address

lukehoffmann commented:
During the course of @benowen-bom's review, I've also gone through the exercise of testing the plugins in a raincal-like workflow. This has resulted in a highly simplified version of the raincal training scripts, which I've added to this repo [Forecast-Improvement/raincal_lite](https://github.com/Forecast-Improvement/raincal_lite).

During this process, I changed the design of the plugins slightly. They now accept the same type of "`model_config_dict`" object as the `ApplyRainForestsCalibration` plugins. This specifies the full set of lead times and thresholds, and the full paths for all of the trained and compiled model files.

After this PR is merged, we will need further changes to provide the acceptance tests and documentation for these plugins. I've intentionally kept those out of this PR to limit the amount of changes to be reviewed at once, and to allow the design to continue to change as needed during this review.

benowen-bom commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -30,7 +30,10 @@ def __init__(
         Args:
             model_config_dict:
-                Dictionary containing Rainforests model configuration variables.
+                Dictionary describing the high-level RainForests model structure;
+                - top level key describes the lead-hour,
+                - next level key describes the threshold,
+                - corresponding values locate the associated model file.
```
Thanks for updating this! Much clearer now.

benowen-bom commented on improver/calibration/rainforest_training.py:
```diff
@@ -89,11 +94,11 @@ def __init__(
-        # Merge default params with optional params.
-        lightgbm_params = lightgbm_params or {}
-        self.lightgbm_params = self.lightgbm_params | lightgbm_params
+        # Merge default params with overrides from constructor argument.
+        if lightgbm_params:
+            self.params = self.params | lightgbm_params
```
Thanks for updating this. I know the change is subtle but this is much easier to follow.

benowen-bom reviewed (dismissed):
Thanks @lukehoffmann for working through my suggestions. I've had a look over the changes and I'm happy to approve the PR now, particularly the additions to the tests. I appreciate that the nature of these routines make it tricky to test, but I think where this has landed is a good balance between thorough without being excessive.

Great work! I think it would be good to have another review (I see that you have requested a review from @dmentipl).

benowen-bom commented:
Also, I think @2384 address the failing check above.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+        model_config_dict: dict[int, dict[str, dict[str, str]]],
+        toolchain="gcc",
+        verbose=False,
+        parallel_comp=0,
```
These are missing type annotations.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+                (e.g. 'gcc-7').
+            verbose (bool):
+                Print verbose output during compilation
+            parallel_comp (int):
```
We don't use type annotations in the docstring as they should be in the function signature.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+        self.verbose = verbose
+        self.treelight_params = {"parallel_comp": parallel_comp, "quantize": 1}
+
+    def process(self, allow_missing=False):
```
These are missing type annotations.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+        """Compile all configured LightGBM models with Treelite.
+
+        Args:
+            allow_missing (bool):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+                    Path(threshold["treelite_model"]),
+                )
+
+    def _compile_model(self, lightgbm_path, output_path):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,142 @@
+        """Compile a lightgbm model with Treelite.
+
+        Args:
+            lightgbm_path (pathlib.Path):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+        training_data,
+        observation_column,
+        training_columns,
+        lightgbm_params=None,
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+    ):
+        """Initialise the options used when training models.
+
+        Args:
+            model_config_dict:
+                Dictionary describing the high-level RainForests model structure;
+                - top level key describes the lead-hour,
+                - next level key describes the threshold,
+                - corresponding values locate the associated model file.
+            training_data (pandas.DataFrame):
+                Combined data set used to train models.
+            observation_column (str):
+                The column in the data set to be trained for.
+            training_columns (List(str)):
+                Set of columns from the data set to be trained from.
+            lightgbm_params (Dict):
+                Optional. Parameters passed into training library. Any parameters
+                here will override the default parameters.
```
Same as above.

lukehoffmann replied:
Not sure why it's not showing up here, but I've fixed this too

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+            thresholds (list of str):
+                Threshold values for which the observation column is trained.
+                Formatted to match the keys used in the model_config object.
+        """
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+
+        Returns:
+            The model object (lightgbm.Booster)
+        """
```
Same as above.

dmentipl reviewed (changes requested):
Thanks @lukehoffmann. This looks good. I've just got some comments about type annotations.

lukehoffmann committed:
Fix type hints

lukehoffmann committed:
Fix imports

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,141 @@
+        toolchain: str = "gcc",
+        verbose: bool = False,
+        parallel_comp: int = 0,
+    ):
```
It's missing the return type here and for the other functions/methods.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,141 @@
+        self.verbose = verbose
+        self.treelight_params = {"parallel_comp": parallel_comp, "quantize": 1}
+
+    def process(self, allow_missing: bool = False):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_compilation.py:
```diff
@@ -0,0 +1,141 @@
+                    Path(threshold["treelite_model"]),
+                )
+
+    def _compile_model(self, lightgbm_path: pathlib.Path, output_path: pathlib.Path):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+        "seed": 0,
+    }
+
+    def __init__(
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+        if lightgbm_params:
+            self.params = self.params | lightgbm_params
+
+    def process(self, lead_time: int, thresholds: list[str]):
```
Same as above.

dmentipl commented on improver/calibration/rainforest_training.py:
```diff
@@ -0,0 +1,150 @@
+            model_path = Path(self.config[lead_time][threshold]["lightgbm_model"])
+            self._train_model(float(threshold), model_path)
+
+    def _train_model(self, threshold: float, model_path: pathlib.Path):
```
Same as above.

dmentipl reviewed (changes requested):
Thanks @lukehoffmann. It's just missing the return types.

lukehoffmann committed:
Type hints for return types

dmentipl reviewed (dismissed):
Thanks @lukehoffmann.

lukehoffmann committed:
Merge branch 'master' into rainforests_training

lukehoffmann committed:
Remove second email
